content/articles/haskell-websocket-tsung-benchmark.md

```diff
@@ -0,0 +1,275 @@
+for reasons I didn't have time to debug.
+
+But without the distributed testing there's a problem: A single server
+can only open ~65000 connections because ports are limited to 16
+bits. We want more connections though!
+
+Luckily tsung supports using multiple virtual IP addresses for a
+single network interface out of the box. We went to Amazon and clicked
+"Assign new IP" a few times to assign more private IPs to our tsung
+box.
+
+Now we associate the new IPs with our network interface:
+
+```shell
+ip addr add 172.31.23.115/20 dev eth0
+ip addr add 172.31.23.113/20 dev eth0
+ip addr add 172.31.23.114/20 dev eth0
+ip addr add 172.31.23.112/20 dev eth0
+ip addr add 172.31.18.80/20 dev eth0
+ip addr add 172.31.18.81/20 dev eth0
+ip addr add 172.31.18.82/20 dev eth0
+ip addr add 172.31.18.83/20 dev eth0
+```
+
+We have a
+[slightly different](https://github.com/WeAreWizards/haskell-websockets-tsung-benchmark/blob/master/code/src/tsung-conf.xml)
+tsung config from the Phoenix people which we copy to our tsung box:
+
+```console
+$ nixops scp --to tsung-1 code/src/tsung-conf.xml tsung-conf.xml
+code/src/tsung-conf.xml -> root@52.31.104.126:tsung-conf.xml
+tsung-conf.xml                                                  100% 1494     1.5KB/s   00:00
+```
+
+# Running tsung
+
+Our Nix config tuned the TCP stack and increased kernel limits, but we
+still need to run `ulimit` to make sure we're not hitting the default
+limit of 1024 file descriptors:
+
+```text
+$ nixops ssh tsung-1
+[root@tsung-1:~]# ulimit -n 2000000
+[root@tsung-1:~]# tsung -f tsung-conf.xml start
+Starting Tsung
+Log directory is: /root/.tsung/log/20151104-1622
+```
+
+tsung exports some data via a web interface on port 8091. We use an
+extra SSH tunnel so we can access this data on http://127.0.0.1:8091:
+
+```console
+$ ssh root@tsung-1 -L 8091:127.0.0.1:8091
+```
+
+## Problem 1: The firewall
+
+All our Nix boxes are configured with a firewall enabled. This is
+because I start from a template configuration instead of starting from
+scratch.
+
+The firewall uses connection tracking to make decisions, and
+connection tracking requires memory. When that memory is full the
+`dmesg` logs look like this:
+
+```
+[ 2960.570157] nf_conntrack: table full, dropping packet
+[ 2960.575060] nf_conntrack: table full, dropping packet
+[ 2960.629764] nf_conntrack: table full, dropping packet
+[ 2960.678016] nf_conntrack: table full, dropping packet
+[ 2992.936177] TCP: request_sock_TCP: Possible SYN flooding on port 8080. Sending cookies.  Check SNMP counters.
+[ 2998.005969] net_ratelimit: 364 callbacks suppressed
+```
+
+That log also shows that we triggered the kernel's DOS protection
+against SYN flooding. We fixed that by increasing
+`net.ipv4.tcp_max_syn_backlog` and `net.core.somaxconn`.
+
+Now when running tsung we got up to about 100k connections on the
+Haskell websocket box:
+
+```
+[root@websock-server:~]# netstat -ntp  | grep -v TIME_WAIT | wc
+ 119748  838238 12094489
```

Comment:
I changed it to `wc -l` which counts lines and only prints one number. That's less ambiguous.